Add config to run minions on GCE using Ubuntu.

andyzheng0831 opened the pull request:
It is for running nodes on Ubuntu, and the work for running master
on Ubuntu will be added later. The configuration consists of several
upstart jobs, which is passed to node instances through GCE metadata
and parsed by cloud-init.

Commit:
Add config to run minions on GCE using Ubuntu.

It is for running nodes on Ubuntu image upto 14.04 LTS (Trusty).
The change for running master on Ubuntu will be added later.
The configuration consists of several upstart jobs, which is
passed to node instances through GCE metadata and parsed by cloud-init.

andyzheng0831 commented:
cc/ @dchen1107
cc/ @yinghan
cc/ @alex-mohr 
cc/ @roberthbailey

cjcullen commented:
ok to test

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+Content-Type: multipart/mixed; boundary="===============6024533374511606659=="
+MIME-Version: 1.0
+
+--===============6024533374511606659==
```
debugging?

andyzheng0831 replied:
No, it is not for debugging. The configuration consists of multiple upstart jobs, and we need to pack them into one file and pass it to GCE metadata (the metadata key is "user-data" as shown in cluster/gce/ubuntu/helper.sh). Ubuntu cloud-init provides a tool https://github.com/number5/cloud-init/blob/master/tools/make-mime.py to pack upstart jobs into an MIME file. So, this header and boundary string are generated by that tool. Cloud-init then can parse the yaml file and extract each upstart job.

dchen1107 replied:
Is MIME format required for ubunbu? what's the benefits?

andyzheng0831 replied:
It is not required by ubuntu, but a quite useful tool. The major benefit of MIME is packing different formats and multiple configuration files into one file. This is important in GCE. We pass the configuration file as "user-data", which does not support multiple files.

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+From nobody Tue Aug 11 10:13:54 2015
```
Is below auto-generated?

dchen1107 replied:
I am referring to line 1 - line 9 here.

andyzheng0831 replied:
Generated by the tool https://github.com/number5/cloud-init/blob/master/tools/make-mime.py for packing multiple upstart jobs into one MIME file.

dchen1107 replied:
ACK.

zmarano commented:
Ubuntu 15.04+ as well as Debian 8+ are systemd based and no longer include or use upstart. You should do one of the following:
- Limit the scope of this change to Ubuntu versions up to 14.10 or even just 14.04 LTS (Trusty) - and likewise Debian 7.
- Detect if systemd is installed and do nothing (also add a comment) until systemd scripts and support is added.
- Or just add a comment in the usage saying systemd is not yet supported and that this won't work on the latest Ubuntu/Debian versions.

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+Content-Transfer-Encoding: 7bit
+Content-Disposition: attachment; filename="kube-addons.conf"
+
+#upstart-job
```
Should I expect all upstart jobs described here are running sequentially?

andyzheng0831 replied:
They are not completely sequentially executed. The dependency is defined in the line "start on" of each job. The whole sequence is as follows:
(1) kube-env.conf and kube-install-packages.conf (they are in parallel);
(2) kube-install-minion.conf is after kube-env.conf as it sources the env variables there;
(3) kubelet.conf and kube-proxy are after kube-install-packages.conf and kube-install-minion.conf. They are in parallel;
(4) kube-docker.conf is after kubelet.conf;
(5) kube-addons.conf is after kube-docker.conf

andyzheng0831 commented:
@zmarano 
Thanks for the information! I will rename it to "trusty" for simplicity. And I will add a README and put the image version notice there. Detecting systemd requires add new logic added to the current config, because none of these upstart jobs will be executed on ubuntu 15.04+.

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+	. /etc/kube-env
+	if [ "${ENABLE_NODE_LOGGING}" = "true" ]; then
+		if [ "${LOGGING_DESTINATION}" = "gcp" ]; then
+			cp /etc/saltbase/kubernetes/saltbase/salt/fluentd-gcp/fluentd-gcp.yaml /etc/kubernetes/manifests/
```
There is no salt installed on those nodes, right? why still created those directories?

andyzheng0831 replied:
Correct, no salt installed. The purpose of pulling salt tarball is for supporting addons. The addon yaml files are in cluster/saltbase/salt. One way is to upload needed yaml files to GCE metadata server and then downloaded by node instances. But this way is hard to extend. If we have new kubelet addons, we have to change both the shell and node.yaml. Another way is like here, just simply upload the entire saltbase tarball, and nodes can use any yaml file. When we add new kubelet addons, we can simply change this particular upstart job to place the yaml to the right directory.

dchen1107 replied:
Yeah, I realized this when I read the below cloud_config file. I did the same thing for coreos demo, I think.

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+script
+	. /etc/kube-env
+	# Assemble docker deamon options
+	echo "DOCKER_OPTS=\"-p /var/run/docker.pid ${EXTRA_DOCKER_OPTS} --log-level=\"debug\" --bridge cbr0 --iptables=false --ip-masq=false\"" > /etc/default/docker
```
Should I assume one version of kubelet is already running here, so that it can do those network configuration work?

dchen1107 replied:
Ok, the clause "start on started kubelet" does make sure kubelet is started here.

andyzheng0831 replied:
Correct, "start on started kubelet" means it is triggered only after kubelet.conf job is successfully started (i.e., kubelet process is up). But kubelet is up does not imply cbr0 is created. So, I added a loop to check cbr0 to avoid restarting docker daemon too early.

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+	done
+	initctl restart docker
+	# Remove docker0
+	ifconfig docker0 down
```
If kubelet is running and it has responsibility to config cbr0, docker0 should be removed by that kubelet too?

andyzheng0831 replied:
I don't see where kubelet code will delete docker0. In my experiments, I also saw docker0 existing after cbr0 was created. If this logic is implemented in kubelet, we should remove these two lines from node.yaml

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+	initctl restart docker
+	# Remove docker0
+	ifconfig docker0 down
+	brctl delbr docker0
```
ditto....

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+	short_hostname=$(hostname -s)
+	hostname $short_hostname
+
+	# We have seen that GCE image may have strict host firewall rules which drop most inbound/forwarded packets. In such a case, add rules to accept all TCP/UDP packets.
```
cc/ @thockin and @ArtfulCoder for sanity-check on this part.

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,281 @@
+	# Create required directories.
+	mkdir -p /var/lib/kubelet
+	mkdir -p /var/lib/kube-proxy
+	mkdir -p /etc/kubernetes/manifests
```
You already refer to this directory in above defined upstart job called: "start on stopped kube-docker". Shouldn't that job be failed?

dchen1107 replied:
Ok. I realized that this is one of first upstart jobs run.

andyzheng0831 replied:
The kube-docker.conf job just sets the command line in /etc/default. If later docker daemon crashes or fails, there is another system upstart job docker.conf will be responsible for restarting it, and it will pick the rewritten command line in /etc/default

dchen1107 commented:
Ok, Can we or-order the upstart jobs defined in this file based on the dependecies? I know you define them through clause "start on", but it is hard to check the correctness of the orders. I basically have to extract all clauses with "start on" and check the dependencies first, then go to each job definition to check job itself. It make the maintenance harder too.

andyzheng0831 commented:
@dchen1107 
I will reorder the jobs in node.yaml in their execution orders.

dchen1107 commented:
Both shippable and travis failed. Can you run hack/update-generated-docs.sh?

andyzheng0831 commented:
No such hack/update-generated-docs.sh file. Do you mean hack/verify-gendocs.sh, or any other file?

dchen1107 commented:
you need to sync your branch to HEAD in this case.

dchen1107 commented:
In this pr, there is no healthcheck on kubelet, kube-proxy and docker, right?

andyzheng0831 commented:
Only pid monitoring. If docker, kubelet, or kube-proxy process is down, upstart will restart it. As for more deeper monitoring such as the ones used in current monit/supervisor, this version does not have it. But we will add it shortly.

dchen1107 commented:
Add a TODO for this then. 

LGTM in general. cc/ @zmerlynn

andyzheng0831 commented:
Added the TODO in node.yaml, and also corrected the username of TODO in helper.sh. The test results will be pasted when it is done.

Thank you all for the comments!

dchen1107 commented:
LGTM. Thanks!

andyzheng0831 commented:
Sorry for not pasting e2e test results promptly. When I run it on GCE, it is disturbed by the reboot test. After running "ifconfig eth0 down/up", ssh to the instances never works unless reboot them. This makes all tests the reboot test failed. I will first revert the change to reboot.go and rerun all tests.

andyzheng0831 commented:
Here is the e2e test results:
Setting:
export KUBE_OS_DISTRIBUTION=trusty
export KUBE_GCE_MINION_IMAGE=ubuntu-1404-trusty-v20150805
export KUBE_GCE_MINION_PROJECT=ubuntu-os-cloud

Summarizing 5 Failures:
[Fail] ReplicationController [It] should serve a basic image on each replica with a private image
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/rc.go:127
(This is a reproducible failure)

[Fail] Examples e2e [Example]ClusterDns [It] should create pod that uses dns
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/examples.go:524
(This test should be in the skipped test list)

[Fail] Kubectl client Simple pod [It] should support port-forward
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:201
(This is a reproducible failure)

[Fail] kube-ui [It] should check that the kube-ui instance is alive
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kube-ui.go:44
(It seems that kube-ui is not enabled in test. After sshing to master (running on Debian) I don't see kube-ui yaml under /etc/kubernetes/addon.)

[Fail] Services [It] should work after restarting kube-proxy
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:308
(This failure is because the testing code restart kube-proxy use init.d, but in Ubuntu we should use initctl restart kube-proxy.)

Ran 112 of 142 Specs in 6496.815 seconds
FAIL! -- 107 Passed | 5 Failed | 2 Pending | 28 Skipped

dchen1107 commented:
Looked into all failed cases together with @andyzheng0831, here are the current status:

[Fail] ReplicationController [It] should serve a basic image on each replica with a private image

This pr installs latest docker release: 1.8.X, which has known issue for image name contains underscore: #12506

[Fail] Examples e2e [Example]ClusterDns [It] should create pod that uses dns

This test should be skipped when testing 1.0.X release. It should passed with HEAD.

[Fail] kube-ui [It] should check that the kube-ui instance is alive

This is test code issue. We should fix it in test. I filed a separate issue for this. 
This should be a real issue in your pr. 

[Fail] Services [It] should work after restarting kube-proxy

I filed a separate issue for test: #12689

andyzheng0831 commented:
For ReplicationController test, I revise node.yaml to install docker 1.7.1 instead of the latest 1.8.1, and the test passes.

For kube-ui, I omitted adding this env-variable in kube-env. After fixing it, test passes.

I will upload the revised commit shortly

andyzheng0831 commented:
Oops, seems I updated a wrong version from an incorrect local branch. Will correct it now.

dchen1107 commented:
LGTM Thanks for looking into this.

roberthbailey commented on cluster/gce/trusty/node.yaml:
```diff
@@ -0,0 +1,286 @@
+clusters:
+- name: local
+  cluster:
+    insecure-skip-tls-verify: true
```
If you aren't going to do it in this PR, please file an issue to make this secure (and cc @stephenR).

andyzheng0831 replied:
Thanks for the comments. This PR is not a complete solution. It provides a version with major functionality and pass e2e tests. This will be quite helpful for unblocking our other work. So, I will make several followup PRs to fix various problems and enrich the functionality. Can we still make change to this PR after lgtm label was added? If prefer to leave fixes in the next PR. Is that OK?

roberthbailey replied:
Fixing it in follow-up PRs is fine with me; just wanted to make sure it was going to get fixed.

andyzheng0831 replied:
As per request from oncall, I fix these problems in this PR to get merged. I change it to the same as configure-vm.sh

roberthbailey commented on cluster/gce/trusty/node.yaml:
```diff
@@ -0,0 +1,286 @@
+users:
+- name: kubelet
+  user:
+    token: ${KUBELET_TOKEN}
```
On debian, we've switched to using client certs for kubelets instead of tokens (kube-proxy still uses tokens though).

andyzheng0831 replied:
Fixed, will upload the new version soon.

roberthbailey commented on cluster/gce/trusty/node.yaml:
```diff
@@ -0,0 +1,286 @@
+
+script
+	. /etc/kube-env
+	# If kubelet or kube-proxy is not stalled in the image, pull release binaries and put them in /usr/bin.
```
nit: s/stalled/installed

andyzheng0831 replied:
Fixed, will upload the new version soon.

roberthbailey commented on cluster/gce/trusty/node.yaml:
```diff
@@ -0,0 +1,286 @@
+		rm "/tmp/${k8s_sha1}"
+	fi
+
+	# Put saltbase configuration files in /etc/saltbase. We will use the add-on yaml files.
```
It's unfortunate that we need to download all of the saltbase files on a non-salt system configuration. You should file an issue (cc @fgrzadkowski and @davidopp) to figure out how to break this out.

andyzheng0831 replied:
Kubelet currently has just one add-on yaml file, so another lightweight way is just upload that particular file to GCE metadata server. This at least satisfies the current usage. If people wish to figure out another way, I think we should also consider the master. Master uses much more add-on yaml files. Uploading each of them to GCE metadata looks not a wise approach. Either just pack the needed files for speed, or use whatever we currently have (i.e., saltbase tarball) for simplicity.

roberthbailey commented on cluster/gce/trusty/node.yaml:
```diff
@@ -0,0 +1,286 @@
+		--config=/etc/kubernetes/manifests \
+		--allow_privileged=false \
+		--v=2 \
+		--cluster_dns=10.0.0.10 \
```
cluster dns shouldn't be hard-coded. It needs to be set from the kube env.

andyzheng0831 replied:
Fixed, will upload the new version soon.

roberthbailey commented on cluster/gce/trusty/node.yaml:
```diff
@@ -0,0 +1,286 @@
+		--allow_privileged=false \
+		--v=2 \
+		--cluster_dns=10.0.0.10 \
+		--cluster_domain=cluster.local \
```
same with cluster domain.

andyzheng0831 replied:
Fixed, will upload the new version soon.

eparis replied:
All of these flag usages with `_` broke the build....    kubelet flags should use `-`

andyzheng0831 replied:
Is this applied to all kubelet flags? The salt-based configuration https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kubelet/default still uses "--api_servers=".

zmerlynn replied:
In one place, that looks like a bug.

eparis replied:
I actually already merged a 'fix' and apologize if I messed something up for you.  Yes, all kubelet flags.  I'll check out your link.

andyzheng0831 replied:
Thanks for notifying me. I will upload a PR to fix some settings in node.yaml. I will fix this in it, and cc you.

eparis replied:
side note: I actually have another PR already open to fix the mistake you found   :)

roberthbailey commented on cluster/gce/trusty/node.yaml:
```diff
@@ -0,0 +1,286 @@
+		echo "Sleep 1 second to wait for cbr0"
+		sleep 1
+	done
+	initctl restart docker
```
@cjcullen Any idea if the kubelet will try to configure docker on an upstart (ubuntu trusty) system? It seems silly to do it here and in the kubelet.

andyzheng0831 replied:
That is true. If kubelet can do this, we don't need a system configuration job for it. It means that we can take this upstart job as a temporary solution, if there is a plan to configure docker in kubelet code. Please note that, the current kubelet code for restarting docker daemon does not work. It only considers debian init system and systemd.

roberthbailey commented:
This LGTM since @andyzheng0831 has promised to fix my concerns in a follow-up PR. I'd like to get this in so that we can start looking at getting Jenkins set up to ensure that it doesn't start to bit rot.
